### DISCUSSION TITLE
Axisymmetric model creation in 3D

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#21366] [Q&A General] [answered]
Dear all,
As part of my recent work concerning radiation damage models, i am interested in creating a 3D cylindrical pipe model (see below) that mainly simulate elasticity effects. The damage resulting from irradiation induced swelling is then captured through the thermal strains calculated from `ComputeThermalExpansionEigenstrain`. 
![image](https://user-images.githubusercontent.com/40111479/174796328-65d7e3b2-5308-416c-bb00-ffcb98c19494.png)
I am currently looking at a suitable 3D axisymmetric model from the repository that will form the basis for my model. As far as i checked most of them were reduced 2D models with `coord_type` set to RZ. I wonder how would this command look like if i want to use a 3D model. I am aware a 2D model would suffice for most of the cases, unless if one want to address problems relating to stress effects in the vicinity of the circumference e.g., torsion. My intention here is to compare with analytical results for which a 3D FE model would result in a better comparision.
Kind regards,
Arun

**Comment by GiudGiud** [comment]
Hello

We currently do not have a mesh generator that will extrude a 2D RZ geometry to a 3D model.
Is your geometry extruded in the axial direction? I see to see a bend at the top and at the bottom of this picture. 

You can use free meshing software gmsh or cubit-learn to generate this mesh it does not seem too complicated. 

Guillaume

**Reply by abarun22** [reply]
Well, the issue is not about meshing, but how do we make use of this 3D mesh and run an axi-symmetric model. Below here is my mesh done in cubit and ready to go in to simulation. My concern here is what should be set to this parameter `coord_type` when we want to run this simulation with a cylindrical coordinate system and problem dimensionality as 3. Will `RZ` work even for this case?
![image](https://user-images.githubusercontent.com/40111479/174831105-3f23088c-f0cb-449f-bfb6-96734806c46d.png)

**Reply by GiudGiud** [reply]
if you have a 3D mesh then you should run XYZ coordinates. 
We do not have RthetaZ in MOOSE, only RZ

**Reply by abarun22** [reply]
Yes that's fine. I have gases flowing inside the pipe and so wanted to apply gas pressure in the inner walls. I am wondering how to intertpret radial direction working from within the cartesian system.

**Reply by GiudGiud** [reply]
The radial direction is not treated any particular way in the cartesian solve.
Are your kernels or your material properties dependent on r ?

**Reply by abarun22** [reply]
As said earlier, my boundary condition for gas pressure depends on the radial direction `R`. If i work in cartesian system and choose one of the axes for applying this pressure BC, i will be applying in the wrong direction. Either i would need to choose cylindrical system, which for this 3D case is not possible or i may need to revert to a 2D model.

**Reply by GiudGiud** [reply]
You will have to apply the BC on both X and Y axis, with a projection based on the local angle.

**Reply by abarun22** [reply]
This seems a quite different way. How do one can calculate the local angle and where this would be accounted in the boundary conditions?

**Reply by GiudGiud** [reply]
I think you may want to look at Pressure.C in the tensor mechanics module src/bcs folder. It uses the normal to compute the contribution of the external pressure to the residual on the bc

**Comment by srinath-chakravarthy** [comment]
At least for small deformations, I suppose you can use the local normal
vector to resolve the normal pressure into x and y components (requires a
little bit of coordinate transformation). You might have to write code for
it.

**Reply by abarun22** [reply]
Thanks Srinath for the suggestion. The tricky thing here is the calculation of the local angle. I wanted to clarify what this angle is and how this can be interpreted from the perspective of a FE domain. Any ideas/suggestions here would help me to solve this issue and focus next on integrating these transformed corrdinates in to the pressure BC application module swiftly.

**Comment by srinath-chakravarthy** [comment]
You should not need the local angle, moose. Boundary normals should be
accessible easily _normals[_qp]

**Reply by abarun22** [reply]
With the normals, i am wondering how to calculate the angle theta needed for coordinate conversion as stated here.
![image](https://user-images.githubusercontent.com/40111479/175326744-8d6692e4-5d64-4107-8fa6-091c96e14fb4.png)
You might probably thinking about a different way to solve this problem

**Comment by srinath-chakravarthy** [comment]
You can do it your way and calculating the angle should be trivial right ?
n.x = mag(n)*max(x)* cos(theta), where n is the normal vector and x is the
unit x vector. I think the normals in moose are unit magnitude and so is
the unit vector in x. So the angle to x axis would simply be cos-1(x
component of the normal vector).
In a more general sense, with this procedure, you should be able to apply
an arbitrary normal pressure on any arbitrary plane. The only tricky part
might be to add residuals to the individual components in moose, but you
can use an action to do this i think.

**Reply by abarun22** [reply]
Thanks Srinath for the inputs. With this conversion based approach, theta eventually can be calculated as a inverse cosine of x component of the normal vector. I am now looking for a suitable template to start my developments. I would like to ask if its more relevant to copy the class `Pressure` that applies pressure on boundaries?

**Comment by srinath-chakravarthy** [comment, marked as answer]
I think you should look at pressure action and just modify the scaling
factors to the correct values. Just saying if pressure was a scalar then
p*n is the vector so the components of the vector give you the relevant
values in each direction.

**Reply by abarun22** [reply]
I guess this has already been implemented in Pressure.C. Here is the `computeQpResidual` function from this file that will tell you how the pressure value is resolved in to component directions. I do not know what `_test` does in this case, but this block would for sure apply the pressure in the normal direction on any given surface, which is what we need and this should work correctly without going for actual conversion to cylindrical coordinate systems. 

```
Real
Pressure::computeQpResidual()
{
  return computeFactor() * (_normals[_qp](_component) * _test[_i][_qp]);
}
```
Probably you might want to say something like this:
```
[BCs]
  [./Pressure]
    [./pbc]
      boundary = 'inner'
      factor= 50
      displacements = 'disp_x disp_y'
    [../]
  [../]
[]
```
Here the x & y displacements are given just to specify that the first two resolved components of pressure are meant to be applied on to this boundary 'inner'. This is what i thought based on our earlier discussions, probably you could pour in your suggestions.
Kind regards,
Arun

**Reply by srinath-chakravarthy** [reply]
Yes this is all you need would be my guess. This is a relatively new addition, since in the previous avatar of Pressure BC, we needed to specify a component, see ADPressure.C for the AD version of the old Pressure.C. I can see now it is deprecated and also very recently (in the last 2 days). So all you need to specify is the pressure value for each of your x and y directions and this should do the rest. The _test is necessary from the weak form of the bc. You can easily verify that the correct pressure is being applied by validating with the axisymmetric 2d model and a cross-section in your 3d model.

**Reply by abarun22** [reply]
Hi Srinath, FYI, the new pressure BC works fine with the results comparing ok with the equivalent analytical results. Thanks for the suggestions.
